### ISSUE TITLE
[FEATURE] Add support for custom html scripts for link surveys

### ISSUE THREAD

**Issue by mattinannt** [formbricks/formbricks#2567]
### Is your feature request related to a problem? Please describe.

Users would like to easily include Google Tag Manager / Google Analytics or Facebook / LinkedIn Pixels into their Link surveys.
Currently there is no way to do this.

### Describe the solution you'd like

Allow custom HTML code injection (CSS, javascript,..) into the Link Survey pages. This should be done on a product level in the Look & Feel Settings (`!IS_FORMBRICKS_CLOUD`).

### Additional context

We should only allow this on self-hosted Formbricks instances since otherwise this imposes data privacy issues on Formbricks Cloud since we would need to update the data privacy policy for different link surveys and maybe even include cookie banners for this. To avoid this, we only allow this on self-hosting Formbricks instances.

**Comment by chinmoykalita** [comment]
Hey, @mattinannt, I would like to work on this if @jobenjada has not started on it yet.

**Comment by pierrejo** [comment]
@chinmoykalita I'm very interested in this feature. Do you have any idea when it might be released?

**Comment by jobenjada** [comment]
@pierrejo it me blocking this due to a lack of design capacity. Can you make a rough mock up of where you'd like to handle this (I suppose you've been doing this with other tools before).

I'm happy to review it, write it up in a ticket and hand it over to @chinmoykalita 🙌 

let's get the team to work! 💪

**Comment by pierrejo** [comment]
Thanks for your answer @jobenjada 

I think the best place for that is in the style settings of the form:
![design_form](https://github.com/user-attachments/assets/7973f05e-7c00-4a56-893c-3058ea9548cc)

I need to be able to add Matomo script tags to inject cookies to users.

**Comment by jobenjada** [comment]
hey mate! I don't think it makes sense to add this to the styling settings 🙈 I'll write up a ticket for the upcoming community hackathon in October 🚀

**Comment by pierrejo** [comment]
> hey mate! I don't think it makes sense to add this to the styling settings 🙈 I'll write up a ticket for the upcoming community hackathon in October 🚀 @jobenjada 

You're right, it's better to put it in settings because it doesn't necessarily affect the style of the form. :
![Présentation sans titre (15)](https://github.com/user-attachments/assets/d66c24d6-3893-4cf4-a4f6-d9af63f8ffbb)

**Comment by jobenjada** [comment]
yeah makes a lot more sense :) can you consult with GPT if we need some form of sanitization to prevent this being exploited?

**Comment by pierrejo** [comment]
@jobenjada Although this part is normally in the hands of the admin or the team, there are several ways to secure this part. Here's chatGPT's answer:

---
To allow users to add custom tags in the "Head" section of your form application while ensuring security, you need to implement several security measures. Here are some steps and recommendations to achieve this:

### 1. Validation and Sanitization
Ensure that the tags added by users are properly validated and sanitized before being inserted into the HTML document.

- **Whitelisting**: Allow only specific tags and attributes that you deem safe. For example, you can permit `<script>`, `<meta>`, `<link>`, etc., but with limited attributes.
- **Escaping special characters**: Escape special characters to prevent malicious code injection.

### 2. Using Security Libraries
Use proven security libraries to validate and sanitize user inputs. For example, in JavaScript, you can use libraries like DOMPurify.

```javascript
const DOMPurify = createDOMPurify(window);
const cleanHTML = DOMPurify.sanitize(userInput);
```

### 3. Content Security Policy (CSP)
Implement a Content Security Policy (CSP) to restrict the sources of scripts and other resources that your application can load. This can help prevent XSS (Cross-Site Scripting) attacks.

```html
<meta http-equiv="Content-Security-Policy" content="default-src 'self'; script-src 'self' https://example.com;">
```

### 4. Sandboxing
If possible, use sandboxing techniques to execute scripts in an isolated environment. This can limit potential damage in case of exploitation.

### 5. Monitoring and Auditing
Set up monitoring and auditing mechanisms to detect and respond quickly to suspicious behaviors. Use security monitoring tools to analyze logs and detect anomalies.

### 6. Limiting Privileges
Limit the privileges of users to add custom tags. For example, only administrators or trusted users should have this capability.

### 7. Security Testing
Conduct regular security tests, including penetration tests, to identify and fix potential vulnerabilities.

### Example Code
Here's an example of JavaScript code to validate and sanitize custom tags before inserting them into the HTML document:

```javascript
function sanitizeAndInsertHeadTags(userInput) {
    const DOMPurify = createDOMPurify(window);
    const cleanHTML = DOMPurify.sanitize(userInput);

    // Insert the sanitized tags into the <head> section
    const head = document.head;
    head.insertAdjacentHTML('beforeend', cleanHTML);
}

// Example usage
const userInput = '<script src="https://example.com/analytics.js"></script>';
sanitizeAndInsertHeadTags(userInput);
```

---

**Comment by Syiana** [comment]
Hello,

this feature is a must have! 

the best way would be a tab named "integrations" 

something like this:
<img width="1225" alt="image" src="https://github.com/user-attachments/assets/3d2e0204-3846-4d04-9910-cec6ff2c6b90">

**Comment by Caroga** [comment]
May I suggest including plausible in this list? It's a self-hosted, privacy first, alternative tot Google Analytics. 
Tracking would be enabled by including the following snippet to the form:

```html
<script defer data-domain="domain.tld" src="https://plausible.domain.tld/js/script.js"></script>
``` 

Would this make sense?



> Hello,
> 
> this feature is a must have!
> 
> the best way would be a tab named "integrations" for each survey
> 
> something like this: <img alt="image" width="1225" src="https://private-user-images.githubusercontent.com/36953982/373796490-3d2e0204-3846-4d04-9910-cec6ff2c6b90.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDE3MDkwNzUsIm5iZiI6MTc0MTcwODc3NSwicGF0aCI6Ii8zNjk1Mzk4Mi8zNzM3OTY0OTAtM2QyZTAyMDQtMzg0Ni00ZDA0LTk5MTAtY2VjNmZmMmM2YjkwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzExVDE1NTkzNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTg3MjdjNTAwNmQxYTQ4NjNlZjdkZjk5NjBkMTIwNzhiZWQxZTEzOGY4OWM4YjM4MTU4MTQyYTk1MWI4MmFkNjMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.lMmjRfoEBs5-JPgjxAhSjCMs26LGP5Y-YLICThhNUvM">

**Comment by Syiana** [comment]
how is this still missing?

**Comment by cameocreative2019** [comment]
Is this actually assigned to another person?

**Comment by arnaudbagnis** [comment]
up

**Comment by jobenjada** [comment]
@arnaudbagnis hey mate! Do you need the code snippet to be the same for each survey / survey link? Or does this need to be customizable per survey?

**Comment by arnaudbagnis** [comment]
Good question 🙂

From my point of view, it should ideally support both levels, but if I had to prioritize:

1) Per survey / survey link (most important)
Each survey can have different needs:
- different analytics tools (GA, Plausible, Matomo, Meta pixel…)
- different domains / data-domain values
- different clients or products
So having the snippet configurable per survey is essential.

2) Optional global default (nice to have)
A global default snippet at product or workspace level could be useful as a fallback,
but it should be overridable at the survey level.

My main use case is:
- self-hosted Formbricks
- link surveys embedded via iframe or standalone
- need for proper analytics & conversion tracking (UX, funnels, replay tools, etc.)

Without per-survey customization, it becomes hard to use Formbricks in multi-client or multi-project setups.

Happy to clarify further or help test once there’s an implementation 🙂

**Comment by jobenjada** [comment]
@arnaudbagnis perfect, that maps with my initial idea of building it. Let me see what we can do!
